pyccl/tests/test_power.py

```diff
@@ -187,89 +205,278 @@ def test_power_sigma8norm_norms_consistent(tf):
+                   'delta_matter_x_delta_matter': -pk_arr})
+
+    pk_CCL_input = -ccl.power.nonlin_matter_power(cosmo_input, k_arr, 0.5)
+
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
+    # Test again passing only linear Pk, but letting HALOFIT do its thing
+    kl_arr = np.logspace(-4, 1, 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, kl_arr, a)
+                        for a in a_arr])
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        pk_linear={'a': a_arr, 'k': kl_arr,
+                   'delta_matter_x_delta_matter': pkl_arr},
+        use_halofit=True)
+
+    pk_CCL_input = ccl.power.nonlin_matter_power(cosmo_input, k_arr, 0.5)
+
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
+    # Test extra power spectrum
+    kl_arr = np.logspace(-4, 1, 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, kl_arr, a)
+                        for a in a_arr])
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        pk_linear={'a': a_arr, 'k': kl_arr,
+                   'delta_matter_x_delta_matter': pkl_arr,
+                   'a_x_b': pkl_arr},
+        pk_nonlin={'a': a_arr, 'k': k_arr,
+                   'delta_matter_x_delta_matter': -pk_arr},
+        use_halofit=True)
+
+    pk_CCL_input = cosmo_input._pk_nl['a_x_b'].eval(k_arr, 0.5, cosmo_input)
+
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
+    # Via `nonlin_power`
+    pk_CCL_input = ccl.power.nonlin_power(cosmo_input, k_arr, 0.5,
+                                          p_of_k_a='a_x_b')
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
 
 def test_input_nonlin_raises():
-    cosmo_input = ccl.Cosmology(Omega_c=0.27, Omega_b=0.05, h=0.7, n_s=0.965,
-                                A_s=2e-9)
+    cosmo = ccl.Cosmology(Omega_c=0.27, Omega_b=0.05, h=0.7,
+                          n_s=0.965, sigma8=0.8,
+                          transfer_function='bbks')
+    a_arr = np.linspace(0.1, 1.0, 50)
+    k_arr = np.logspace(np.log10(2e-4), np.log10(1), 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, k_arr, a)
+                        for a in a_arr])
+    pk_arr = np.array([ccl.power.nonlin_matter_power(cosmo, k_arr, a)
+                       for a in a_arr])
+
+    # Not a dictionary
+    with pytest.raises(TypeError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin=np.pi)
+
+    # k not present
+    with pytest.raises(ValueError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin={'a': a_arr, 'kk': k_arr,
+                       'delta_matter_y_delta_matter': pk_arr})
+
+    # delta_matter_x_delta_matter not present
```

Comment:
Does the calculator always have to have the matter power spectrum?

Reply:
See my comment/questions about the API below. The idea is that this needs to eventually be present, so if `use_halofit = False`, `pk_nonlin` should contain `delta_matter_x_delta_matter`.